benchmark-framework/src/main/java/io/openmessaging/benchmark/WorkloadGenerator.java

```diff
@@ -436,53 +444,77 @@ private TestResult generateLoad(List<BenchmarkProducer> producers, long testDura
                     rateFormat.format(consumeRate), throughputFormat.format(consumeThroughput),
                     dec.format(currentBacklog / 1000.0), //
-                    dec.format(reportHistogram.getMean() / 1000.0),
-                    dec.format(reportHistogram.getValueAtPercentile(50) / 1000.0),
-                    dec.format(reportHistogram.getValueAtPercentile(99) / 1000.0),
-                    dec.format(reportHistogram.getValueAtPercentile(99.9) / 1000.0),
-                    throughputFormat.format(reportHistogram.getMaxValue() / 1000.0));
+                    dec.format(publishReportHistogram.getMean() / 1000.0),
+                    dec.format(publishReportHistogram.getValueAtPercentile(50) / 1000.0),
+                    dec.format(publishReportHistogram.getValueAtPercentile(99) / 1000.0),
+                    dec.format(publishReportHistogram.getValueAtPercentile(99.9) / 1000.0),
+                    throughputFormat.format(publishReportHistogram.getMaxValue() / 1000.0));
 
             result.publishRate.add(publishRate);
             result.consumeRate.add(consumeRate);
             result.backlog.add(currentBacklog);
-            result.publishLatencyAvg.add(reportHistogram.getMean() / 1000.0);
-            result.publishLatency50pct.add(reportHistogram.getValueAtPercentile(50) / 1000.0);
-            result.publishLatency75pct.add(reportHistogram.getValueAtPercentile(75) / 1000.0);
-            result.publishLatency95pct.add(reportHistogram.getValueAtPercentile(95) / 1000.0);
-            result.publishLatency99pct.add(reportHistogram.getValueAtPercentile(99) / 1000.0);
-            result.publishLatency999pct.add(reportHistogram.getValueAtPercentile(99.9) / 1000.0);
-            result.publishLatency9999pct.add(reportHistogram.getValueAtPercentile(99.99) / 1000.0);
-            result.publishLatencyMax.add(reportHistogram.getMaxValue() / 1000.0);
-
-            reportHistogram.reset();
+            result.publishLatencyAvg.add(publishReportHistogram.getMean() / 1000.0);
+            result.publishLatency50pct.add(publishReportHistogram.getValueAtPercentile(50) / 1000.0);
+            result.publishLatency75pct.add(publishReportHistogram.getValueAtPercentile(75) / 1000.0);
+            result.publishLatency95pct.add(publishReportHistogram.getValueAtPercentile(95) / 1000.0);
+            result.publishLatency99pct.add(publishReportHistogram.getValueAtPercentile(99) / 1000.0);
+            result.publishLatency999pct.add(publishReportHistogram.getValueAtPercentile(99.9) / 1000.0);
+            result.publishLatency9999pct.add(publishReportHistogram.getValueAtPercentile(99.99) / 1000.0);
+            result.publishLatencyMax.add(publishReportHistogram.getMaxValue() / 1000.0);
+
+            result.endToEndLatencyAvg.add(endToEndReportHistogram.getMean());
+            result.endToEndLatency50pct.add(endToEndReportHistogram.getValueAtPercentile(50));
```

Comment:
Don't have to divide these by 1000.0 like above because these are in millis, while publish latency is in nanos?

Reply:
Pub latency is collected in micros (we have nanoseconds but that's too much), while for end-to-end we only have millis.

Reply:
Nitpick, but would be nice to have the same unit all-around if possible :).

Reply:
Yes, good point. I was thinking in terms of efficiency.. but probably overthinking here.